### PR TITLE
[FEATURE] V3 - Implement column_pair helpers/providers for SqlAlchemyExecutionEngine

### PR DESCRIPTION
**Pull request by NathanFarmer**
Changes proposed in this pull request:
- JIRA: GREAT-56/GREAT-122
- JIRA: GREAT-56/GREAT-126

The helper functions that are registered in this PR are what was needed to get the `ColumnPairValuesEqual` metric working which will come in a later PR.

### Definition of Done
Please delete options that are not relevant.

- [X] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/en/latest/contributing/style_guide.html?highlight=style%20guide)
- [X] I have performed a [self-review](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html?highlight=checklist) of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have run any local integration tests and made sure that nothing is broken

### CODE REVIEW

File: great_expectations/self_check/util.py [hunk 1]
```diff
@@ -1118,7 +1118,7 @@ def candidate_test_is_on_temporary_notimplemented_list(context, expectation_type
             "expect_column_bootstrapped_ks_test_p_value_to_be_greater_than",
             # "expect_column_kl_divergence_to_be_less_than",
             "expect_column_parameterized_distribution_ks_test_p_value_to_be_greater_than",
-            "expect_column_pair_values_to_be_equal",
+            # "expect_column_pair_values_to_be_equal",
```
Comment (alexsherstinsky) [inline great_expectations/self_check/util.py]: @NathanFarmer This is a V2 section.  We should leave it unchanged.  Thanks.

File: great_expectations/self_check/util.py [hunk 2]
```diff
@@ -1239,7 +1239,7 @@ def candidate_test_is_on_temporary_notimplemented_list_cfe(context, expectation_
             # "expect_column_min_to_be_between",
             # "expect_column_sum_to_be_between",
             "expect_column_pair_values_A_to_be_greater_than_B",
-            "expect_column_pair_values_to_be_equal",
+            # "expect_column_pair_values_to_be_equal",
```
Comment (alexsherstinsky) [inline great_expectations/self_check/util.py]: @NathanFarmer Correct (the "CFE" section is V3).  Thanks.

File: great_expectations/expectations/metrics/map_metric_provider.py [hunk 3]
```diff
@@ -580,6 +580,69 @@ def inner_func(
 
         return wrapper
 
+    elif issubclass(engine, SqlAlchemyExecutionEngine):
```
Comment (alexsherstinsky) [inline great_expectations/expectations/metrics/map_metric_provider.py:583]: 🙇

File: great_expectations/expectations/metrics/column_pair_map_metrics/column_pair_values_equal.py [hunk 4]
```diff
@@ -11,11 +15,19 @@ class ColumnPairValuesEqual(ColumnPairMapMetricProvider):
...
-    # TODO: <Alex>ALEX -- temporarily only a Pandas implementation is provided (others to follow).</Alex>
+    # TODO: <Alex>ALEX -- temporarily only Pandas and SQL Alchemy implementations are provided (Spark to follow).</Alex>
     @column_pair_condition_partial(engine=PandasExecutionEngine)
     def _pandas(cls, column_A, column_B, **kwargs):
         ignore_row_if = kwargs.get("ignore_row_if")
         if not ignore_row_if:
             ignore_row_if = "both_values_are_missing"
 
         return column_A == column_B
+
```
Comment (alexsherstinsky) [inline great_expectations/expectations/metrics/column_pair_map_metrics/column_pair_values_equal.py:22]: 🙇 🚀

File: great_expectations/expectations/metrics/column_pair_map_metrics/column_pair_values_equal.py [hunk 5]
```diff
@@ -11,11 +15,19 @@ class ColumnPairValuesEqual(ColumnPairMapMetricProvider):
     domain_keys = ("batch_id", "table", "column_A", "column_B")
     default_kwarg_values = {"ignore_row_if": "both_values_are_missing"}
 
-    # TODO: <Alex>ALEX -- temporarily only a Pandas implementation is provided (others to follow).</Alex>
+    # TODO: <Alex>ALEX -- temporarily only Pandas and SQL Alchemy implementations are provided (Spark to follow).</Alex>
     @column_pair_condition_partial(engine=PandasExecutionEngine)
     def _pandas(cls, column_A, column_B, **kwargs):
         ignore_row_if = kwargs.get("ignore_row_if")
```
Comment (alexsherstinsky) [inline great_expectations/expectations/metrics/column_pair_map_metrics/column_pair_values_equal.py]: @NathanFarmer Do we really need the `ignore_row_if` processing here?  It should come from the expectations layer.  Let us delete it from here and similar Metrics-level objects.  Thanks!

Comment (NathanFarmer) [reply great_expectations/expectations/metrics/column_pair_map_metrics/column_pair_values_equal.py]: @alexsherstinsky I agree. I think this is a relic of the old way of handling that argument.

File: great_expectations/expectations/metrics/map_metric_provider.py [hunk 6]
```diff
@@ -1637,6 +1774,83 @@ def _sqlalchemy_column_map_condition_values(
...
+    metrics: Dict[str, Any],
+    **kwargs,
+):
+    """Return values from the specified domain that match the map-style metric in the metrics dictionary."""
+    (
+        boolean_mapped_unexpected_values,
+        compute_domain_kwargs,
+        accessor_domain_kwargs,
+    ) = metrics["unexpected_condition"]
+
```
Comment (alexsherstinsky) [inline great_expectations/expectations/metrics/map_metric_provider.py]: @NathanFarmer Minor: Let us delete this line (to keep up the existing style).  Thanks!

File: great_expectations/expectations/metrics/map_metric_provider.py [hunk 7]
```diff
@@ -1637,6 +1774,83 @@ def _sqlalchemy_column_map_condition_values(
...
+        sa.select(sa.column(column_A_name), sa.column(column_B_name))
+        .select_from(selectable)
+        .where(boolean_mapped_unexpected_values)
+    )
+
+    result_format = metric_value_kwargs["result_format"]
+    if result_format["result_format"] != "COMPLETE":
+        query = query.limit(result_format["partial_unexpected_count"])
+
+    return [dict(val) for val in execution_engine.engine.execute(query).fetchall()]
```
Comment (alexsherstinsky) [inline great_expectations/expectations/metrics/map_metric_provider.py]: @NathanFarmer This does not seem correct.  For `COLUMN_PAIR`, we should not be getting the column names.  Please see the output of the `Pandas` version.

**Review by alexsherstinsky — CHANGES_REQUESTED** [review]
@NathanFarmer This is great -- and appears to be close!  Please see some comments.  Thanks!

**Review by alexsherstinsky — APPROVED** [review]

**Pull request merged**